frappe/model/document.py

```diff
@@ -848,9 +848,7 @@ def check_docstatus_transition(self, to_docstatus):
 				self._action = "submit"
 				self.check_permission("submit")
 			elif self.docstatus.is_cancelled():
-				raise frappe.DocstatusTransitionError(
-					_("Cannot change docstatus from 0 (Draft) to 2 (Cancelled)")
-				)
+				self.check_permission("cancel")
```

Comment:
This is invalid. This change makes current feature work but breaks other validations. 

Also `draft->discard` should just require `write` instead of `cancel`. Cancel perm is for cancelling submitted docs.